vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+				},
+
+				"algorithm": {
+					Type:        framework.TypeString,
+					Description: "Signing algorithm to use. This will default to RS256, and is currently the only allowed value.",
+					Default:     "RS256",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.CreateOperation: i.pathOIDCCreateUpdateKey,
+				logical.UpdateOperation: i.pathOIDCCreateUpdateKey,
+				logical.ReadOperation:   i.pathOIDCReadKey,
+				logical.DeleteOperation: i.pathOIDCDeleteKey,
+			},
+			ExistenceCheck:  i.pathOIDCKeyExistenceCheck,
+			HelpSynopsis:    "oidc/key/:key help synopsis here",
+			HelpDescription: "oidc/key/:key help description here",
+		},
+		{
+			Pattern: "oidc/key/" + framework.GenericNameRegex("name") + "/rotate/?$",
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the key",
+				},
+				"verification_ttl": {
+					Type:        framework.TypeDurationSecond,
+					Description: "Controls how long the public portion of a key will be available for verification after being rotated. Setting verification_ttl here will override the verification_ttl set on the key.",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: i.pathOIDCRotateKey,
+			},
+			HelpSynopsis:    "oidc/key/:key/rotate help synopsis here",
+			HelpDescription: "oidc/key/:key/rotate help description here",
+		},
+		{
+			Pattern: "oidc/key/?$",
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ListOperation: i.pathOIDCListKey,
+			},
+			HelpSynopsis:    "list oidc/key/ help synopsis here",
+			HelpDescription: "list oidc/key/ help description here",
+		},
+		{
+			Pattern: "oidc/.well-known/openid-configuration/?$",
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ReadOperation: i.pathOIDCDiscovery,
+			},
+			HelpSynopsis:    "read oidc/.well-known/discovery/ help synopsis here",
+			HelpDescription: "read oidc/.well-known/discovery/ help description here",
+		},
+		{
+			Pattern: "oidc/.well-known/keys/?$",
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ReadOperation: i.pathOIDCReadPublicKeys,
+			},
+			HelpSynopsis:    "read oidc/.well-known/keys/ help synopsis here",
+			HelpDescription: "read oidc/.well-known/keys/ help description here",
+		},
+		{
+			Pattern: "oidc/token/" + framework.GenericNameRegex("name"),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the role",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ReadOperation: i.pathOIDCGenerateToken,
+			},
+		},
+		{
+			Pattern: "oidc/role/" + framework.GenericNameRegex("name"),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the role",
+				},
+				"key": {
```

Comment:
In my memory of past discussions we'd decided to do the same thing we do in the combined database backend where the keys would list the roles allowed to use those keys so that someone that can create roles doesn't automatically get key access as well if they don't have permission to update that key configuration. Is this planned, or scrapped, or...?

Reply:
I don't recall that being discussed for this feature, but it wouldn't be hard. You mean an `allowed_roles` parameter on the key, and role create/updates would check that?

Reply:
Added a jira

Reply:
FWIW at some point yesterday I was talking to BK and he actually thought this was in the RFC. It's not, but it's good practice. And yes, an `allowed_roles` which could support globbing including `*`.

In a future where we may support cross-namespace generation this also sets up a pattern where we can add something like `allowed_external_roleids`.